### ISSUE TITLE
Deadlocks using gRPC C# in Unity - Possible shutdown race - processing GOAWAY after shutdown

### ISSUE THREAD

**Issue by rib** [grpc/grpc#21633]
### What version of gRPC and what language are you using?
C#

grpc_unity_package.2.27.0-dev.zip - build ID 44bbb740-e047-4210-8aea-0ab70d40ccb6

### What operating system (Linux, Windows,...) and version?
Windows 10

### What runtime / compiler are you using (e.g. python version or version of gcc)
Unity 2019.2

### What did you do?
I'm trying to investigate a deadlock in Grpc.Core related to the Environment shutdown. The deadlock seems to happen in many cases after having used the Grpc API, after I switch out of Unity's play mode and then later try and re-enter play mode Unity freezes. Another way of triggering the deadlock is to build a standalone application and the deadlock will manifest when I try and quit the application.

My gRPC client code in Unity is interacting with a game server written in Rust ([Tonic](https://github.com/hyperium/tonic)) which seems to work fine when I test using a headless gRPC client written in Go.

I have since tried to come up with a minimal reproduction and found one example that seems to be a fairly reliable reproduction with a server streaming request where I dispose the stream without reading anything from it.

<details><summary>This is the protocol I'm testing with</summary>

```protobuf
syntax = "proto3";

package test.grpc;

message Empty {}

message Ping {
    uint64 epochSendTimestampNS = 1;
}

message Pong {
    uint64 epochSendTimestampNS = 1;
}

service Test {
    rpc OpenTestStream(Empty) returns (stream Empty);
    rpc SendPing(Ping) returns (Pong);
}
```
</details>

<details><summary>This is the C# test function I have</summary>

```cs
public async void RunGrpcTest()
    {
        var certUri = Path.Combine(Application.streamingAssetsPath, "ca.crt");
        var req = UnityWebRequest.Get(certUri);
        await req.SendWebRequest();
        var cacert = req.downloadHandler.text;

        var cancellationTokenSource = new CancellationTokenSource();

        var interceptor = new AsyncAuthInterceptor(async (context, metadata) =>
        {
            // dummy token, with a "username" claim
            metadata.Add("authorization", "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6InRlc3QifQ.hNQI_r8BATy1LyXPr6Zuo9X_V0kSED8ngcqQ6G-WV5w");
        });

        string address = "" + testAddress + ":" + testPort;
        ChannelCredentials credentials = null;

        if (useTls)
        {
            credentials = new SslCredentials(cacert);
            credentials = ChannelCredentials.Create(new SslCredentials(cacert),
                CallCredentials.FromInterceptor(interceptor));
        }
        else
        {
            credentials = ChannelCredentials.Insecure;
        }

        var channel = new Channel(address, credentials);

        var client = new Test.Grpc.Test.TestClient(channel);

        try {
            var sentTs = GetEpochTimestampNS();
            var pong = await client.SendPingAsync(new Test.Grpc.Ping { EpochSendTimestampNS = sentTs },
                                                    deadline: DateTime.UtcNow.AddSeconds(10),
                                                    cancellationToken: cancellationTokenSource.Token);
            var receivedTs = pong.EpochSendTimestampNS;
            var delta = receivedTs - sentTs;
            Debug.Log("Ping sent to Frontend (timestamp = " + sentTs + "),  Pong Received (timestamp = " + receivedTs + "), Round Trip = " + delta);

            Debug.Log("Opening test stream...");
            using (var streamCall = client.OpenTestStream(new Test.Grpc.Empty {},
                                                          new CallOptions(deadline: DateTime.UtcNow.AddSeconds(30),
                                                                          cancellationToken: cancellationTokenSource.Token)))
            {
                // With a Rust/Tonic server then if we don't read anything from the stream we tend to get a deadlock
#if false
                if (await streamCall.ResponseStream.MoveNext()) {
                    Debug.Log("Received Empty{} item on test stream");
                } else {
                    Debug.Log("Nothing read from test stream");
                }
#else
                Debug.Log("DEBUG: Immediately disposing create group stream");
#endif
            }
        }
        finally
        {
            Debug.Log("Shutting down channel...");
            await channel.ShutdownAsync();
        }

        Debug.Log("Done");
    }
```

</details>

TL;DR the test sends one unary `SendPing` request to the server followed by a server streaming `OpenTestStream` request but doesn't wait to read anything from the stream before disposing the stream and shutting down the channel.

**Looking at the logs from `GRPC_TRACE="api,channel,connectivity_state,executor"`; something that seems to reliably correlate with test runs that will lead to a deadlock is seeing that a GOAWAY is received _after_ the Environment has shut everything down. (after tracing  `grpc\src\core\lib\surface\init.cc:212: grpc_shutdown(void)`)**

I don't know yet, but it seems surprising to see any log messages related to network I/O after the grpc state has been shutdown, so I wonder if maybe there is some kind of reference counting problem and something should be blocking the shutdown from completing while there is apparently still a connection to the server that's open (or some buffered data that needs to be flushed somehow)?

This is a trace log for a deadlock case:

  https://pastebin.com/7YZ94k0d

In this case I was able to run the test four times successfully, but on the fifth run I see this extra information after shutting down grpc (search for `"grpc_shutdown(void)"` in the log):
```
INFO: 2106225329392 C:\Users\Robert\src\grpc\src\core\ext\transport\chttp2\transport\chttp2_transport.cc:1088: ipv4:192.168.0.21:50051: Got goaway [1] err={"created":"@1578765853.378000000","description":"GOAWAY received","file":"C:\Users\Robert\src\grpc\src\core\ext\transport\chttp2\transport\chttp2_transport.cc","file_line":1077,"grpc_status":14,"http2_error":1,"raw_bytes":""}
INFO: 2106225329392 C:\Users\Robert\src\grpc\src\core\lib\transport\connectivity_state.cc:151: ConnectivityStateTracker client_transport[000001EA602B84D0]: READY -> TRANSIENT_FAILURE (got_goaway)
INFO: 2106225329392 C:\Users\Robert\src\grpc\src\core\lib\transport\connectivity_state.cc:158: ConnectivityStateTracker client_transport[000001EA602B84D0]: notifying watcher 000001EA56953890: READY -> TRANSIENT_FAILURE
INFO: 2106225329392 C:\Users\Robert\src\grpc\src\core\lib\transport\connectivity_state.cc:78: watcher 000001EA56953890: delivering async notification for TRANSIENT_FAILURE
INFO: 2106225329392 C:\Users\Robert\src\grpc\src\core\lib\transport\connectivity_state.cc:151: ConnectivityStateTracker client_channel[000001EA57C68950]: READY -> IDLE (helper)
INFO: 2106225329392 C:\Users\Robert\src\grpc\src\core\lib\transport\connectivity_state.cc:151: ConnectivityStateTracker client_transport[000001EA602B84D0]: TRANSIENT_FAILURE -> SHUTDOWN (close_transport)
INFO: 2106225329392 C:\Users\Robert\src\grpc\src\core\lib\transport\connectivity_state.cc:158: ConnectivityStateTracker client_transport[000001EA602B84D0]: notifying watcher 000001EA56953890: TRANSIENT_FAILURE -> SHUTDOWN
INFO: 2106225329392 C:\Users\Robert\src\grpc\src\core\lib\transport\connectivity_state.cc:78: watcher 000001EA56953890: delivering async notification for SHUTDOWN
INFO: 0 C:\Users\Robert\src\grpc\src\core\lib\surface\call.cc:573: grpc_call_unref(c=000001EA655903E0)
```

This is another trace log for reference that doesn't trigger a deadlock with an equivalent server written in Go:

https://pastebin.com/xcASsX2F

_Update: for a while I thought I wasn't able to reproduce this with the Go-based server, but actually I can. It looks like this issue is timing sensitive and changes to the test (such as enabling/disabling TLS) that probably just change timing will affect how reliably this issue is reproduced_

At this point, I'm not familiar with the Grpc.Core code to know how to try and fix this, so I'm really hoping someone here will recognise what's probably wrong based on the log.

If it would be helpful, I can aim to push a git repo with a minimal Unity test and the Rust + Go servers I've used to reproduce this?

If you need me to rerun the test with additional logging information, please let me know.

For me this issue is a rather serious blocker for progressing my project, so I'll continue trying to investigate myself too but any help from someone familiar with the Grpc.Core implementation would be hugely appreciated!

I haven't figured out how to rebuild the unity packages for the grpc libraries but have at least been able to rebuild Grpc.Core and grpc_csharp_ext.dll so I can start to experiment with local changes.

**Comment by rib** [comment]
Something I've just found that helps me be able to reproduce this more reliably, is following these steps:

1. Run the gRPC server (either Rust or Go based server will lead to a deadlock this way...)
2. Enter Unity play mode
3. Run the test gRPC requests
4. Exit Unity play mode
_(At this point the logs may look clean, and I wouldn't predict a deadlock - no GOAWAY message)_
E.g. last messages will be:
```
src\core\lib\surface\init.cc:212: grpc_shutdown(void)
Done
src\core\lib\surface\call.cc:573: grpc_call_unref(c=000001A811B2B120)
```
5. Shutdown the server (doesn't have to be done immediately - e.g. waiting a few seconds)
Wait a second and I will then see these trace messages:
```
src\core\lib\transport\connectivity_state.cc:151: ConnectivityStateTracker client_transport[000001A823C92B90]: READY -> SHUTDOWN (close_transport)
src\core\lib\transport\connectivity_state.cc:158: ConnectivityStateTracker client_transport[000001A823C92B90]: notifying watcher 000001A823C463C0: READY -> SHUTDOWN
src\core\lib\transport\connectivity_state.cc:78: watcher 000001A823C463C0: delivering async notification for SHUTDOWN
src\core\lib\transport\connectivity_state.cc:151: ConnectivityStateTracker client_channel[000001A8238892B0]: READY -> IDLE (helper)
```

Now try to re-enter play mode and it hits a deadlock - this seems to be 100% reliable for me so far.

**Comment by rib** [comment]
From testing some more variations, I'm now finding that testing against the Go server I trigger deadlocks without needing to shut the server down - it's just that I'm running without TLS now (which I _suspect_ is really just changing the timing of things).

In this case I'm not actually seeing any spurious looking trace messages after the grpc_shutdown, so maybe those are going to be a red herring or I'm hitting a slightly different deadlock here.

**Comment by rib** [comment]
To help reproduce this I've pushed a [repo here](https://github.com/rib/grpc-unity-deadlock-reproduction) which includes a minimal Unity project and two test servers (one written in Rust, one in Go).

For reference I'm using Unity 2019.2.17f1 (though have reproduced this on earlier 2019.2 versions too)

The Go server can be run like:
```
cd test-server-go
go run main.go
```
or with TLS enabled (I've included self-signed certs in the repo for testing):
```
go run main.go -use_tls=true
```

The Rust server can be run like:
```
cd test-server-rust
RUST_LOG=trace cargo run
```
or with TLS:
```
RUST_LOG=trace USE_TLS=1 cargo run
```

The servers will listen on port `50051` by default.

In the Unity project if you open `TestScene` there is a `TestCase` object which has a few configurable options (e.g. for enabling TLS) and when entering play mode there will simply be a fullscreen 'Test' button that can be pressed to trigger the test gRPC requests. By default it connects to `127.0.0.1:50051`.

**Comment by rib** [comment]
_btw for building the Rust server, you'll need to need to have the latest 1.40 rust compiler:_

https://www.rust-lang.org/tools/install

`rustup update stable`

**Comment by cg-adam** [comment]
We are getting similar issues here as well. We have a stream we are starting, but it is also blocking ALL other calls through GRPC (cannot perform any unary calls once the stream has started). As well as it blocked, we cannot seem to shutdown/cancel/stop the stream, if we do further calls are still blocked (unary and streams).

**Comment by rib** [comment]
Hi @jtattermusch I was wondering if there is any more information I can provide that might help investigate or reproduce this? Combined with issue #21734 these bindings basically don't seem to be usable with Unity currently :(

I wonder if the problem might somehow stem from the fact that Unity's SynchronizationContext will schedule all tasks on the same (main) thread by default (unless using `ConfigureAwait(false)`) and maybe that's currently breaking some assumption within these bindings?

**Comment by cg-adam** [comment]
@rib Even when trying to use ConfigureAwait(false) we are still getting blocked on all calls.

**Comment by rib** [comment]
Just to mention that I've just tried testing with a few older versions: 2.25 (commit 9b288b20c4e5) and 2.23 (commit e14154481baf7) and get the same deadlock.

Would love to hear from anyone that knows of a version they have managed to use with Unity somehow.

Would hugely appreciate if one of the maintainers could comment on whether they have been able to run the test case I made.

**Comment by tylandercasper** [comment]
Did you ever have any luck resolving this? I'm trying to get Text to speech to work in Unity it completely hangs the editor :(

**Comment by rib** [comment]
Unfortunately not, I basically gave up :(

I've tried multiple version of Unity, I've tried multiple versions of these C# bindings and I've tried multiple async/await integrations for Unity and haven't been able to find a combination that works - I actually don't currently understand how anyone is apparently using these bindings since I really couldn't find any version where the editor wouldn't hang.

I guess the Unity support for this project isn't actively developed or maintained?

Since I'm using Rust for my server-side Grpc service I started looking at binding my own native plugin where I'd just use Rust for the client-side Grpc and made some progress with that, but that's on a back-burner now. In general I had to switch to other things in my project and abandon adding multi-user support for now since I had really been counting on these binding working.

I feel quite gutted that no developer/maintainer for this project was even able to try reproducing the issue by running the test case project I made - considering that it takes a fair amount of effort in itself to create a minimal test case.

At this point I'm concerned that maybe these C# bindings might no longer be developed / maintained now that [grpc-dotnet](https://github.com/grpc/grpc-dotnet) exists, except that's unfortunately of no use to anyone using Unity.

**Comment by EliCDavis** [comment]
@rib I echo your feelings. The DLLs found in their unity package on their build page seems to throw errors as well. I have to manually go through NuGet and grab versions that will work with my unity project. This issue goes across Major version releases of Unity as well.

**Comment by rib** [comment]
@EliCDavis Yeah, never heard any response from a maintainer here so pretty much just decided the Unity support is unmaintained and it would be too much of a liability to depend on in production. I switched to binding in my own plugin written in Rust where I could use https://github.com/hyperium/tonic instead.

Honestly I think I wish they never claimed to support Unity because it led me to waist a huge amount of time which was a big deal for me while I'm trying to boostrap a self-funded startup. Has led me to have general bad feelings about this project - to me it's a huge red flag that it's so difficult to get feedback from maintainers.

**Comment by kyleger** [comment]
Same problem. It makes using gRPC in the Unity Editor impossible. A response to this issue would be greatly appreciated.

**Comment by inaryart** [comment]
> @rib I echo your feelings. The DLLs found in their unity package on their build page seems to throw errors as well. I have to manually go through NuGet and grab versions that will work with my unity project. This issue goes across Major version releases of Unity as well.

The same happened in my C++ project, the client is normal when the exe exits, and the client is deadlocked when the dll exits.
call stack:
```
 	ntdll.dll!NtWaitForAlertByThreadId()	Unknown
 	ntdll.dll!RtlSleepConditionVariableSRW()	Unknown
 	KernelBase.dll!SleepConditionVariableSRW()	Unknown
 	MyProjectPlatform.pyd!absl::lts_20210324::synchronization_internal::Waiter::Wait(class absl::lts_20210324::synchronization_internal::KernelTimeout)	C++
 	MyProjectPlatform.pyd!AbslInternalPerThreadSemWait_lts_20210324()	C++
 	MyProjectPlatform.pyd!absl::lts_20210324::CondVar::WaitCommon(class absl::lts_20210324::Mutex *,class absl::lts_20210324::synchronization_internal::KernelTimeout)	C++
 	MyProjectPlatform.pyd!gpr_cv_wait()	C++
 	MyProjectPlatform.pyd!grpc_timer_manager_shutdown(void)	C++
 	MyProjectPlatform.pyd!grpc_shutdown_internal_locked(void)	C++
 	MyProjectPlatform.pyd!grpc_shutdown()	C++
 	MyProjectPlatform.pyd!grpc::Channel::~Channel(void)	C++
 	MyProjectPlatform.pyd!std::_Ref_count<class grpc::Channel>::_Destroy(void)	C++
>	MyProjectPlatform.pyd!std::_Ref_count_base::_Decref() line 1110	C++
 	MyProjectPlatform.pyd!std::_Ptr_base<grpc::ChannelInterface>::_Decref() line 1341	C++
 	MyProjectPlatform.pyd!std::shared_ptr<grpc::ChannelInterface>::~shared_ptr<grpc::ChannelInterface>() line 1629	C++
 	MyProjectPlatform.pyd!MyProjectCloud::MyProjectCloudInterface::Stub::~Stub()	C++
 	MyProjectPlatform.pyd!MyProjectCloud::MyProjectCloudInterface::Stub::`scalar deleting destructor'(unsigned int)	C++
 	MyProjectPlatform.pyd!std::default_delete<MyProjectCloud::MyProjectCloudInterface::Stub>::operator()(MyProjectCloud::MyProjectCloudInterface::Stub * _Ptr) line 3120	C++
 	MyProjectPlatform.pyd!std::unique_ptr<MyProjectCloud::MyProjectCloudInterface::Stub,std::default_delete<MyProjectCloud::MyProjectCloudInterface::Stub>>::~unique_ptr<MyProjectCloud::MyProjectCloudInterface::Stub,std::default_delete<MyProjectCloud::MyProjectCloudInterface::Stub>>() line 3232	C++

```

grpc version: 1.38.0

**Comment by AntonPetrov83** [comment]
@rib for me the problem was solved when I passed `Channel.ShutdownToken` as a cancellationToken of a streaming method. 
Can you try to use `Channel.ShutdownToken` in your call to `OpenTestStream()`?

**Comment by PlinChen** [comment]
In my opinion, `Task` in C# is to blame. 
Deadlock is always there if `async Task` and `await ` are used widely.
`gRPC for Unity` uses much of `async Task`, and troubles much.

**Comment by vitsensei** [comment]
I'm having some deadlock issue when shutting down as well.
My client is C++ and server in C#. 

My client has its own server where the real "server" can establish a Channel to enable bi-direction communication.

I noticed that if during an RPC from C# to C++, C++ decided to do an RPC to C#, it triggers the condition for hanging during shutdown().

However, I cannot reproduce the problem with a minimal example, so I think there's something else going on here...

I discuss briefly in this Google Group chat: https://groups.google.com/g/grpc-io/c/ly-JwdzS1EM

Seems like there are others encounter this as well.

**Comment by jtattermusch** [comment]
I'm going to close issue this since we're not going to pursuit Unity/Xamarin support in `Grpc.Core` anymore.
- The `Grpc.Core` implementation is now in maintenance mode and we're only going to focus on critical and security fixes.
- Supporting Unity/Xamarin simply presents too much maintenance overhead and it's also difficult to test with the test infrastructure available to us
- The Unity/Xamarin support has always been classified as "experimental", so discontinuing the support shouldn't be too surprising.

See [the recent update on gRPC C#](https://groups.google.com/g/grpc-io/c/OTj5mb1qzb0/m/-M81UQipAQAJ) for more details.

Note that in practice, running Grpc.Core on Unity/Xamarin might still be possible (there is nothing in principle that would prevent it from working), but it will be completely at your own risk and it's simply not one of the use cases the grpc team will help you with as `Grpc.Core` on Unity/Xamarin is now considered as "not supported" by us (but you might still be able to get help from the community).

**Comment by andrekoehler** [comment]
@AntonPetrov83 Thanks to your tip I was able to solve my deadlock. Instead of using `Channel.ShutdownToken` I just used a token from my own CancellationTokenSource which I cancel during Unity's `OnDisable` message. Unity's calls the `OnDisable` message on all game objects whenever the Unity editor stops play mode. So it seems all gRPC streams MUST be cancelled or else gRPC will keep running some thread pool threads even after GrpcEnvironment.ShutdownChannelsAsync was called.
